src/webgpu/gpu_test.ts

```diff
@@ -29,39 +34,136 @@ type TypedArrayBufferViewConstructor =
+        this.holder = undefined;
+      }
       throw ex;
+    } finally {
+      // Mark the holder as free. (This only has an effect if the pool still has the holder.)
+      // This could be done at the top but is done here to guard againt async-races during release.
+      holder.acquired = false;
     }
   }
 
-  async acquire(): Promise<GPUDevice> {
-    assert(this.state !== 'acquired', 'Device was in use');
-    assert(this.state !== 'failed', 'Failed to initialize WebGPU device');
+  // Gets a device and creates a DeviceHolder.
+  // If the device is lost, DeviceHolder.lostReason gets set.
+  private static async makeHolder(): Promise<DeviceHolder> {
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+
+    const holder: DeviceHolder = {
+      acquired: true,
+      device: await adapter.requestDevice(),
+      lostReason: undefined,
+    };
+    holder.device.lost.then(ev => {
+      holder.lostReason = ev.message;
+    });
+    return holder;
+  }
+
+  // Create error scopes that wrap the entire test.
+  private beginErrorScopes(): void {
+    assert(this.holder !== undefined);
+    this.holder.device.pushErrorScope('out-of-memory');
+    this.holder.device.pushErrorScope('validation');
+  }
+
+  // End the whole-test error scopes. Check that there are no extra error scopes, and that no
+  // otherwise-uncaptured errors occurred during the test.
+  private async endErrorScopes(): Promise<void> {
+    assert(this.holder !== undefined);
+    let gpuValidationError: GPUValidationError | GPUOutOfMemoryError | null;
+    let gpuOutOfMemoryError: GPUValidationError | GPUOutOfMemoryError | null;
 
-    const state = this.state;
-    this.state = 'acquired';
-    if (state === 'uninitialized') {
-      await this.initialize();
+    try {
+      // May reject if the device was lost.
+      gpuValidationError = await this.holder.device.popErrorScope();
+      gpuOutOfMemoryError = await this.holder.device.popErrorScope();
+    } catch (ex) {
+      assert(
+        this.holder.lostReason !== undefined,
+        "popErrorScope failed, but device.lost hasn't fired (yet)"
+      );
+      throw ex;
     }
 
-    assert(!!this.device);
-    return this.device;
-  }
+    await assertReject(
+      this.holder.device.popErrorScope(),
+      'There was an extra error scope on the stack after a test'
+    );
 
-  release(device: GPUDevice): void {
-    assert(this.state === 'acquired');
-    assert(device === this.device, 'Released device was the wrong device');
-    this.state = 'free';
+    if (gpuValidationError !== null) {
+      assert(gpuValidationError instanceof GPUValidationError);
+      throw new TestFailedButDeviceReusable(
+        `Unexpected validation error occurred: ${gpuValidationError.message}`
+      );
+    }
+    if (gpuOutOfMemoryError !== null) {
+      assert(gpuOutOfMemoryError instanceof GPUOutOfMemoryError);
+      throw new TestFailedButDeviceReusable('Unexpected out-of-memory error occurred');
```

Comment:
If we had an unexpected OOM, I wonder if it's a good idea to reuse the device as it may cause OOM as well in other unrelated tests.

For some classes of leaks where the Device is holding onto unreachable resources and keeping things alive, it would be better to discard that device. Maybe on unexpected OOMs we can `window.gc` before the next test run. That's outside of the scope of this PR though.

Reply:
Good point; I'll change this error type.

We have attemptGarbageCollection so it would be easy to do that. I'll do it in a followup.